File: app/utils/dawproject/timeline/clips.ts
import { registerTimeline } from "../registry/timelineRegistry";
import type { IClips, ITrack } from "../types";
import { Clip } from "./clip";
import { Timeline } from "./timeline";
import { TimeUnit } from "./timeUnit";

const clipsFactory = (xmlObject: any): Clips => {
  const instance = new Clips();
  instance.fromXmlObject(xmlObject);
  return instance;
};

@registerTimeline("Clips", clipsFactory)
export class Clips extends Timeline implements IClips {
  clips: Clip[];

  constructor(
    clips?: Clip[],
    track?: ITrack,
    timeUnit?: TimeUnit,
    name?: string,
    color?: string,
    comment?: string
  ) {
    super(track, timeUnit, name, color, comment);
    this.clips = clips || [];
  }

  toXmlObject(): any {
    const obj: any = {
      Clips: {
        ...super.toXmlObject(), // get attributes from Timeline
      },
    };

    // Append child elements for each clip
    if (this.clips && this.clips.length > 0) {
      obj.Clips.Clip = this.clips.map(clip => clip.toXmlObject().Clip);
    }

    return obj;
  }

  fromXmlObject(xmlObject: any): this {
    super.fromXmlObject(xmlObject);

    // Process child elements of type Clip
    const clips: Clip[] = [];
    if (xmlObject.Clip) {
      const clipArray = Array.isArray(xmlObject.Clip)
        ? xmlObject.Clip
        : [xmlObject.Clip];
      clipArray.forEach((clipObj: any) => {
        clips.push(new Clip().fromXmlObject(clipObj));
      });
    }
    this.clips = clips;

    return this;
  }
}

File: app/utils/dawproject_python/clips.py
from lxml import etree as ET
from classes.timeline import Timeline
from classes.clip import Clip


class Clips(Timeline):
    XML_TAG = "Clips"

    def __init__(
        self,
        clips=None,
        track=None,
        time_unit=None,
        name=None,
        color=None,
        comment=None,
    ):
        super().__init__(track, time_unit, name, color, comment)
        self.clips = clips if clips else []

    def to_xml(self):
        # Create the XML element for Clips
        clips_elem = ET.Element(self.XML_TAG)

        # Set attributes specific to Clips
        if self.time_unit:
            clips_elem.set("timeUnit", str(self.time_unit))
        if self.track:
            clips_elem.set(
                "track", str(self.track.id)
            )  # Assuming track has an id attribute

        # Append child elements for each clip
        for clip in self.clips:
            clips_elem.append(clip.to_xml())

        return clips_elem

    @classmethod
    def from_xml(cls, element):
        # Initialize instance using the parent class's method
        instance = super().from_xml(element)

        # Process child elements of type Clip
        clips = []
        for clip_elem in element.findall("Clip"):
            clips.append(Clip.from_xml(clip_elem))
        instance.clips = clips

        return instance
